text/0000-prelude-2021.md

```diff
@@ -0,0 +1,172 @@
+
+## `TryFrom`/`TryInto`
+
+The `TryFrom`/`TryInto` traits could not be added to the prelude without an edition boundary. This was [tried](https://github.com/rust-lang/rust/pull/49305) around the time of the 2018 edition, but failed due to breaking substantial parts of the ecosystem. However, `TryFrom`/`TryInto` offer a generic conversion mechanism that is more robust than `as` because the conversions are explicitly fallible. Their usage is more widely applicable than the `From`/`Into` traits because they account for fallibility of the conversions.
+
+Without doing this, the `TryFrom`/`TryInto` traits remain less accessible than the infallible `From`/`Into` conversion traits, providing a disincentive to implement/use fallible conversions even where conversion operations are intrinsically fallible.
+
+## `FromIterator`
+
+The documentation for `FromIterator`'s `from_iter()` method currently reads:
+
+> `FromIterator::from_iter()` is rarely called explicitly, and is instead used through `Iterator::collect()` method. See `Iterator::collect()`'s documentation for more examples.
+
+However, it is reasonably common that type inferencing fails to infer the full type of the target type, in which case an explicit type annotation or turbofishing is needed (such as `iter.collect::<HashMap<_, _>>()` -- the type of the iterator item is available, so wildcards can be used for this). In these cases, `HashMap::from_iter(iter)` can be a more concise and readable way to spell this, which would be easier if the trait was in scope.
+
+## `std::{mem, fmt}`
+
+It is reasonably common to use functions from the `mem` module such as `replace()` or `swap()`. These method names are generic enough that it is useful to keep the `mem::` prefix explicitly. Providing a `mem` name as part of the prelude makes this pattern more obvious and, given the short module name, will not require any imports to make use of the items in this module.
+
+Explicit (non-derive) implementations of `Debug` and `Display` are relatively common and require a set of items from the `fmt` module (the trait itself, the `Formatter` type and the `fmt::Result` type).
+
+## `std::collections::{HashMap, HashSet}`
+
+Especially when coming from another language where hashmaps and sets are language built-ins with literals provided as part of the language, it can be a little surprising at first that these have to be imported before use. Their usage can be especially common in script-like code, for which effortless usage will be especially helpful.
+
+While a number of third-party hashmaps are in common use (most of which reexport or wrap the std `HashMap` with a different hasher), bringing the std `HashMap` into the prelude should not cause issues, since explicit imports will always override prelude imports.
+
+## `std::{{fs::File}, path::{Path, PathBuf}}`
+
+Are very commonly used, not least of which in script-like CLI tools.
+
+## `std::sync::{Arc, Mutex, RwLock}`
```

Comment:
Subjectively, this batch feels a lot less "universal" or "general-purpose" to me than the rest. Multi-threading logic might be contained to a small number of modules, so I feel this is getting into diminishing returns territory. And if these items qualify for the prelude, shouldn’t (insert list of favorites) qualify too?

Overall it’d be nice to have statistics about how often various items are imported in a some corpus of code, to better support the choices made by this RFC in what to include or not. Lacking that, I’d err on the side of fewer additions.